### ISSUE TITLE
Password hash should not rely on system dependant constant

### ISSUE THREAD

**Issue by cypx** [magento/magento2#29235]
### Preconditions (*)

1. Magento 2.3.3
2. PHP 7.2.22

### Steps to reproduce (*)

1. Fresh Magento install
2. Create some admin and customer accounts
3. Make some system or hardware upgrade on server compiling and running PHP which change Sodium predefined constants https://www.php.net/manual/en/sodium.constants.php
4. Try to login with any account

### Expected result (*)

1. Any admin or customer could login with their current password

### Actual result (*)

1. Admin and customer cannot login with their current password and need to use forgotten password function to get a new one

**Temporary workaround**

Get constant value from the old server

```bash
$ php -r 'echo SODIUM_CRYPTO_SIGN_SEEDBYTES."\n";'
32
$ php -r 'echo SODIUM_CRYPTO_PWHASH_OPSLIMIT_INTERACTIVE."\n";'
4
$ php -r 'echo SODIUM_CRYPTO_PWHASH_MEMLIMIT_INTERACTIVE."\n";'
33554432
```

Check if they differ on the new server

```bash
$ php -r 'echo SODIUM_CRYPTO_SIGN_SEEDBYTES."\n";'
32
$ php -r 'echo SODIUM_CRYPTO_PWHASH_OPSLIMIT_INTERACTIVE."\n";'
2
$ php -r 'echo SODIUM_CRYPTO_PWHASH_MEMLIMIT_INTERACTIVE."\n";'
67108864
```

On the new server, replace constant by the value from the old server into the sodium_crypto_pwhash function

https://github.com/magento/magento2/blob/2.4-develop/lib/internal/Magento/Framework/Encryption/Encryptor.php#L585

User could login again without need to reset their password.

---
Please provide [Severity](https://devdocs.magento.com/guides/v2.3/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [x] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

**Comment by cypx** [comment]
This seems confirmed by libsodium author and library example, constant values can change and did change, you must save the function parameters in order to recompute the same value later.

https://github.com/jedisct1/libsodium-php/issues/165#issuecomment-366599786
https://github.com/jedisct1/libsodium-php#encrypt-a-file-using-a-key-derived-from-a-password

**Comment by cypx** [comment]
Can be reproduce using PHP official Docker image, switching from 7.2.19 to 7.2.20 or from 7.3.6 to 7.3.7 must invalidate all the hashes

```bash
$ docker run -it --rm php:7.2.19-cli php -r 'echo SODIUM_CRYPTO_PWHASH_MEMLIMIT_INTERACTIVE."\n";'
33554432
$ docker run -it --rm php:7.2.20-cli php -r 'echo SODIUM_CRYPTO_PWHASH_MEMLIMIT_INTERACTIVE."\n";'
67108864
$ docker run -it --rm php:7.3.6-cli php -r 'echo SODIUM_CRYPTO_PWHASH_MEMLIMIT_INTERACTIVE."\n";'
33554432
$ docker run -it --rm php:7.3.7-cli php -r 'echo SODIUM_CRYPTO_PWHASH_MEMLIMIT_INTERACTIVE."\n";'
67108864
```

**Comment by iphigenie** [comment]
So as people upgrade to 7.3 because of Magento 2.4 they will suddenly have admin login issues?

**Comment by cypx** [comment]
If they upgrade from a Magento version later than 2.3.2, its possible but also depend to previous php environment (the best way to verify is to test constant value before and after).

**Comment by hostep** [comment]
@nathanjosiah: can we have somebody review this, and based on the findings figure out a fix or close the issue?
This sounds like quite an important issue in my opinion.

**Comment by nathanjosiah** [comment]
@magento export issue to Jira project MC as Bug

**Comment by irenelagno** [comment]
@magento export issue to Jira project MC as Bug

**Comment by nathanjosiah** [comment]
Internal issue MC-40854 created for investigation

**Comment by nathanjosiah** [comment]
This issue has been fixed in 2.4.2, 2.4.1-p1, 2.3.7, 2.3.6-p1 and as part of internal tickets MC-36789, MC-36886, MC-36790, and MC-37973 respectively.
